benchmarkcommon/src/test/java/com/ss/benchmark/httpclient/HCPerformanceTests.java

```diff
@@ -0,0 +1,395 @@
+package com.ss.benchmark.httpclient;
+
+import com.codahale.metrics.*;
+import com.ss.benchmark.httpclient.common.Payloads;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.testng.annotations.AfterTest;
+import org.testng.annotations.BeforeTest;
+import org.testng.annotations.Test;
+
+import java.util.HashSet;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+/**
+ * @author sharath.srinivasa
+ */
+@Test(groups = "performance")
+public abstract class HCPerformanceTests {
+
+    protected static final String ECHO_DELAY_BASE_URL = "/echodelayserv/echo";
+    protected static final String ECHO_DELAY_SHORT_URL = "/echodelayserv/echo/short";
+    protected static final String ECHO_DELAY_LONG_URL = "/echodelayserv/echo/long";
+    protected static final String BASE_URL = "http://localhost:8080";
+
+    protected static final int EXECUTIONS = 10_000;
+
+
+    private static final Logger logger = LoggerFactory.getLogger(HCPerformanceTests.class);
+
+    protected final MetricRegistry metricRegistry = new MetricRegistry();
+    protected final ScheduledReporter reporter = ConsoleReporter.forRegistry(metricRegistry).convertDurationsTo(TimeUnit.MILLISECONDS).build();
+
+
+    private Set<CountDownLatch> latches = new HashSet<>();
+
+    private HC client;
+
+    @BeforeTest
+    public void initializeTest() {
+        reporter.start(1, TimeUnit.HOURS);
+
+        client = getClient();
+
+        client.setup(BASE_URL);
+
+        // this is simply to warmup the connection pool
+        logger.info("Start warmup");
+        for (int i = 0; i < EXECUTIONS; i++) {
+            syncGET(metricRegistry.timer(MetricRegistry.name(this.getClass(), "warmup", "timing")),
+                    metricRegistry.counter(MetricRegistry.name(this.getClass(), "warmup", "errorRate")));
+        }
+        logger.info("Completed warmup");
+    }
+
+    @AfterTest
+    public void finalizeTest() {
+        for (CountDownLatch latcher : latches) {
+            try {
+                latcher.await();
+            } catch (InterruptedException e) {
+                e.printStackTrace();
+            }
+        }
+
+        reporter.report();
+        reporter.close();
+    }
+
+    @Test(groups = {"sync", "blocking"})
+    public void testSyncBlockingShortGET() {
+        String method = myName();
+        logger.info("Start " + method);
```

Comment:
I'm not a huge of logging start-end of a function call.  However, you can minimize mistakes with something like this:

```
    public static void main() throws Exception {
        log("some-name", () -> 1);
    }

    private static <T> T log(String name, Callable<T> c) throws Exception {
        logger.info("Starting " + name);
        return c.call();
        logger.info("Ending " + name);
    }
```

No biggie though.

Reply:
let's leave for now